File: src/Wc3FrameExtLua/Fdf/Simple/StatusBar.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')

---@type FdfDataClass
local FdfData = require('FdfData') or error('')

--=======
-- Class
--=======

local FdfSimpleStatusBar = Class.new('FdfSimpleStatusBar', FdfData)
---@class FdfSimpleStatusBar : FdfData
local public = FdfSimpleStatusBar.public
---@class FdfSimpleStatusBarClass : FdfDataClass
local static = FdfSimpleStatusBar.static
---@type FdfSimpleStatusBarClass
local override = FdfSimpleStatusBar.override
local private = {}

--=========
-- Static
--=========

---@param name string
---@param child FdfSimpleStatusBar | nil
---@return FdfSimpleStatusBar
function override.new(name, child)
    isTypeErr(name, 'string', 'name')
    if child then isTypeErr(child, FdfSimpleStatusBar, 'child') end

    local instance = child or Class.allocate(FdfSimpleStatusBar)
    instance = FdfData.new(name, 'SIMPLESTATUSBAR', true, nil, instance)

    return instance
end

--========
-- Public
--========

---@param width number
function public:setWidth(width)
    isTypeErr(self, FdfSimpleStatusBar, 'self')
    isTypeErr(width, 'number', 'width')

    local str_width = tostring(width)
    self:setParameter('Width', str_width)
end

---@param height number
function public:setHeight(height)
    isTypeErr(self, FdfSimpleStatusBar, 'self')
    isTypeErr(height, 'number', 'height')

    local str_height = tostring(height)
    self:setParameter('Height', str_height)
end

---@param tex_file string
function public:setBarTexture(tex_file)
    isTypeErr(self, FdfSimpleStatusBar, 'self')
    isTypeErr(tex_file, 'string', 'tex_file')

    self:setParameter('BarTexture', '\"'..tex_file..'\"')
end

--- Look-Up FileNames in some String table (for example gameinterface)
---@param flag boolean
function public:setDecorateFileNames(flag)
    isTypeErr(self, FdfSimpleStatusBar, 'self')
    isTypeErr(flag, 'boolean', 'flag')
    
    if flag then
        self:setParameter('DecorateFileNames')
    else
        self:removeParameter('DecorateFileNames')
    end
end

--=========
-- Private
--=========

return static

File: src/Fdf/Fdf/Simple/StatusBar.ts
import { Fdf } from "../../Fdf";

export class SimpleStatusBar extends Fdf {
    constructor(name: string){
        super(name, 'SIMPLESTATUSBAR', true)
    }

    public get width(){return this._width}
    public set width(w: number){
        this._setParam('Width', w.toString())
        this._width = w
    }

    public get height(){return this._height}
    public set height(h: number){
        this._setParam('Height', h.toString())
        this._height = h
    }

    public get decorateFileNames(){return this._decorate}
    public set decorateFileNames(flag: boolean){
        if (flag){
            this._setParam('DecorateFileNames')
        } else {
            this._removeParam('DecorateFileNames')
        }
        this._decorate = flag
    }

    public get setAllPoints(){return this._all_points}
    public set setAllPoints(flag: boolean){
        if (flag){
            this._setParam('SetAllPoints')
        } else {
            this._removeParam('SetAllPoints')
        }
        this._all_points = flag
    }

    public get barTexture(){return this._bar_texture}
    public set barTexture(path: string){
        this._bar_texture = path
        this._setParam('BarTexture', '\"' + path + '\"')
    }
    
    private _width: number = -1;
    private _height: number = -1;
    private _decorate: boolean = false;
    private _all_points: boolean = false;
    private _bar_texture: string = '';
}